src/puppetlabs/puppetdb/cli/benchmark.clj

```diff
@@ -191,64 +185,86 @@
     hosts catalog changes in accordance to the users preference)
 
   * Submit the resulting catalog"
-  [{:keys [host lastrun catalog report factset base-url run-interval rand-percentage] :as state} clock]
+  [{:keys [host lastrun catalog report factset] :as state} base-url run-interval rand-percentage clock]
   (if-not (> (- clock lastrun) run-interval)
     state
-    (let [catalog (some-> catalog update-catalog (maybe-tweak-catalog rand-percentage))
-          report (some-> report update-report-run-fields)
-          factset (some-> factset (update-factset rand-percentage))]
+    (let [catalog (some->> catalog (maybe-tweak-catalog rand-percentage))
+          factset (some->> factset (maybe-tweak-factset rand-percentage))]
       ;; Submit the catalog and reports in separate threads, so as to not
       ;; disturb the world-loop and otherwise distort the space-time continuum.
       (when catalog
         (future
           (try
-            (client/submit-catalog base-url 6 (json/generate-string catalog))
+            (->> catalog
+                 update-producer-timestamp
+                 json/generate-string
+                 (client/submit-catalog base-url 6))
             (log/infof "[%s] submitted catalog" host)
             (catch Exception e
               (log/errorf "[%s] failed to submit catalog: %s" host e)))))
       (when report
         (future
           (try
-            (client/submit-report base-url 5 (json/generate-string report))
+            (some->> report
+                     update-report
+                     json/generate-string
+                     (client/submit-report base-url 5))
             (log/infof "[%s] submitted report" host)
             (catch Exception e
               (log/errorf "[%s] failed to submit report: %s" host e)))))
       (when factset
         (future
           (try
-            (client/submit-facts base-url 4 (json/generate-string factset))
+            (->> factset
+                 update-producer-timestamp
+                 json/generate-string
+                 (client/submit-facts base-url 4)) 
             (log/infof "[%s] submitted factset" host)
             (catch Exception e
               (log/errorf "[%s] failed to submit factset: %s" host e)))))
       (assoc state
-        :lastrun clock
-        :catalog catalog))))
+             :lastrun clock
+             :factset factset
+             :catalog catalog))))
```

Comment:
I think this change is appropriate because I think it more closely mirrors real life, but we should verify that it's really what we want because it'll change the way the benchmark tool does factsets. Previously mutations on factsets presented as "aberations" from a baseline, whereas this turns it into more of a continuous mutation where the baseline is never revisited. I think the effect will be less contention on factset submission because previously a mutated factset submission would always be followed by another change relative to the mutated state, whereas now that change will only occur rp% of the time.

Do we have a good reason not to be doing this for reports as well?